### PR TITLE
Multi-queue capture serialisation ordering, respecting fence blocking/signalling

### PR DESCRIPTION
**Pull request by GertyP**
## Description

In short: This makes a capture's linear sequence of events better match the sequence that the GPU actually executes (currently for D3D12 only).  This enables and fixes the `D3D12_Multi_Wait_Before_Signal` test and should fix/improve Renderdoc replaying of multi-queue captures for work that get submitted out of 'execution' order, which is totally fine for applications to do but which caused RD problems.

In more detail, from the main commit message -
```

Instead of serialising captured chunks purely by ascending event ID, which fails
to correctly replay the 'D3D12_Multi_Wait_Before_Signal' test, this walks the
commands for all captured queues and blocks/defers serialising queue commands
that are subsequent to a blocked 'fence wait'.

When queue serialisation reaches a blocked 'wait' it cycles queues to try to
serialise the next unblocked queue command chunks.

It tracks fence signal values within the capture so that we can resume serialising
queue command chunks after a suitably signalled, previously blocked fence 'wait'.

This could result in deadlocking the serialisation if a queue waits on a fence that
is signalled outside the scope of what is captured.  To deal with this, we first
assume that 'waited on' fences that are never seen to be signalled in our capture
records are simply considered signalled so we'll never block on them.  Then, if we
detect a deadlock on any other fence waits, we try to conservatively emulate a
suitable fence signal outside the capture's scope (happens in D3D12_Sharing test).
If all else fails, we resort to the previous event-ID-ordered serialisation.
```

Although the changes are quite isolated, they're not quite trivial, so in anticipation of some feedback and discussion, I thought it best to start this off as a draft PR.

### CODE REVIEW

**Comment by baldurk** [comment]
Thanks for the PR and trying to tackle this! It's a really gnarly problem to try and solve so I'd definitely appreciate if you wanted to put the time into solving this. With that said I don't agree that this is the right direction to go in.

For one thing, although it's not necessarily load-bearing, poking into internally serialised chunks like this is not something I want to do. It's quite brittle and e.g. the current code isn't robust and will break on chunks where the values don't land at those precise offsets - which is possible if you enable callstack capturing. Also I'm not sure if it affects your solution here but as a note, what you are calling and using as event IDs for indexing chunks are definitely _not_ event IDs and they won't even be ordered the same way.

The larger concern I have is that I don't think this re-ordering should be done at capture time, I think it will be better to re-order during replay. This approach has the advantage of fixing the application's bad orderings as near to the source as possible and then 'hiding' it from elsewhere, but it also means we're now lying about the calls back made in a more significant way and doesn't allow us the option of showing the actual CPU submission order.

My preferred solution would be to do presumably what the D3D12 runtime does, and insert another layer for queue operations that we can then insert delays into. Whenever we serialise a queue operation (mostly just submit or sparse table update) if it is blocked by a wait we don't replay it immediately but have it be internally pending until such a time as it's unblocked. This could be done either by somehow bookmarking re-serialising the chunk on replay when we get to it, or else by tokenising/saving the results of the serialise and replaying again.

This does then open the question - how should these events be displayed in the UI. There are two competing orders, the order the commands were given in and the order they execute in. Most likely for consistency we would want the second order, but we'd need some way of showing the user the first order otherwise this could get very confusing.

I would also rather have a robust solution to this that doesn't fall back to unblocking queues indeterminately to regain forward progress or giving up. I think as long as you serialise the completed fence values at capture start time you should have all the information to reconstruct the necessary order, as long as the application itself didn't have any unsatisfied waits.

As one other slightly more minor thing, at the moment we don't serialise direct CPU signals of fences which will be necessary information for this kind of reordering/restructuring.

**Comment by GertyP** [comment]
Thanks for the feedback.

Re. the brittleness of hard offsets for 'peekSignal' and 'peekWait' to inspect fence resource IDs and values from the chunks:  Agreed, although any change affecting the offsets used in these peek functions would be caught and prevented by way of a failing test.  Still, that is a rather remote and indirect way of indicating the affected area.

I'm keen to keep any additional intrusion into the rest of the code as minimal as reasonably possible and as it is, I think it's minimally intrusive;  other ideas for alternatives to simply peek at fence and values all seem to needlessly affect multiple existing systems ... which may ultimately be OK but not while such a low intrusive approach seems within reach.

I wonder if an improvement would be to simply add direct validation of the offsets for fence ID (we never really need to go through the machinery to translate a resource ID to an actual ID3D12Fence instance) and value locations by doing a one-time dummy serialisation of test signal and wait to a temp stream, from which we immediately pull those values back out with 'peekWait' and 'peekSignal' to verify they're looking in the right place.

> what you are calling and using as event IDs for indexing chunks are definitely not event IDs and they won't even be ordered the same way

I don't know what they should be called;  I think I saw the term event ID used in conjunction with (or very nearby) the "int64_t to chunk" map containers that the original serialisation seems to order everything by.  I'm still mainly ordering by this int64_t sequence number so what they're really called doesn't affect anything as far as I'm aware.

> I think it will be better to re-order during replay. This approach has the advantage of fixing the application's bad orderings ... doesn't allow us the option of showing the actual CPU submission order.

This approach/view of things is a little surprising.  I've always thought that the CPU submission order is less important and far less useful than GPU execution order, especially if simply replaying in CPU submission order gives wrong results, which is what we have at the moment.  If the API/GPU can happily execute what's been submitted then there is no such thing as 'bad ordering';  the user's application is not required or even expected to submit queue work in an order that enables the GPU to immediately proceed.  In fact it can be much simpler/cleaner for the application to build up some or all of a frame's multi-queue commands and commandlist work and then submit everything by queue, all in one go.  Applications can (and do) queue up work that they expect to be blocked until released by subsequent GPU (or CPU) fence signalling.

I wonder if, because RD interleaves all queue events into a single list, you find the kind of 'GPU execution' re-ordering that I've done here as a bit jarring and you want it to somehow reflect CPU ordering rather than GPU, but if RD kept a clear separation of the events by queue (e.g. in the event browser, having a tab for each queue's event list) then maybe you'd be more accepting since within any one queue's list of events, there's never any re-ordering from this change;  an Event Browser using a tab (or combo selector) for only ever showing one queue at a time, the CPU submission order always matches the GPU execution order and such a layout and separation would show the same sequence with this 'fence-aware' serialisation as it would with the existing CPU-submission-order serialisation of events.  Viewing sequence of events by queue is what Pix does anyway and is a nicer way to see and think about things (than merging them all into a single list of events) anyway.

> ... insert another layer for queue operations that we can then insert delays into. Whenever we serialise a queue operation (mostly just submit or sparse table update) if it is blocked by a wait we don't replay it immediately but have it be internally pending until such a time as it's unblocked.

I don't have a clear enough picture of where this lives and what data and systems will be affected.  Do you think it could be as self-contained and unintrusive as this change's few functions in d3d12_device.cpp?  Would seem a shame if the job of 'GPU execution ordering'  were to spread out beyond the current small area.

> ... a robust solution to this that doesn't fall back to unblocking queues indeterminately to regain forward progress or giving up.

Yeah, without having pre-capture fence values and also without recording CPU-side fence signalling that may be performed during capture, it has to make some (reasonable) assumptions about the signalling of fences.  These are -
- Fences that we see never get signalled with a value sufficient to release a wait have to be assumed to be sufficiently signalled for the captured sequence to actually fully execute in the first place.
- Fences that we see get signalled through the captured events (which misses out CPU signalling) are assumed to start out in an unsignalled/blocking state.
- If we get to a point where there's deadlock, waiting for a fence, then it must have been signalled behind our back (i.e. by CPU) for the submitted work to actually complete.

So this may leave one or two tricky situations.  For instance:  Fence 'A' actually starts out signalled (with a large enough value), then, when capturing, a queue waits on 'A', which this change pessimistically assumes start out blocking (since it sees a later signal of 'A' elsewhere).  It seems like we could help out here by recording the current value of a fence, for any 'wait'.  Still, this change will notice this lock and eventually signal and progress, just not necessarily in the order that the GPU will have run... Although users/applications generally know what to expect about **true** execution;  that if some work can be consumed/executed on more than one queue simultaneously then they'd understand that work on one can be executed before, during, or after work on the others and that where there is an interdependency, they need to correctly marshal that with fence signalling where appropriate.

That leaves CPU fence signalling that goes on behind our backs as a source for discrepancies... but that's already an existing problem anyway, right?  The existing serialising pays no attention to fence signalling for its replay so can (and does) get it wrong.  We have a GPU fence signalling test that replays incorrectly and we could easily also have the same with some contrived CPU fence signalling cases (e.g. Fence 'A' starts unsignalled/blocking, mutiple queues have a load of work submitted, amongst which, one queue waits on 'A' which is only later CPU signalled).  In fact, this PR's changes would execute that scenario correctly while the existing code would replay it incorrectly.

I'd argue this change is no worse than the existing behaviour and in most cases it's better, so it feels like it would be a shame to bin it because it's not perfect (i.e. the lack of pre-capture [or at least pre-wait] fence value tracking and lack of CPU fence signalling).  Even to defer it until it's perfect (e.g. maybe adding pre-wait fence value recording, CPU fence signalling tracking, and perhaps split events out into separate queues in the UI) feels like much more tricky, intrusive work which seems hard to understand, given the old/current behaviour of no regard for fences... Still, far better to discuss now than to regret if and after it might get submitted :smile

**Comment by baldurk** [comment]
I appreciate what you're saying, and if you're wanting a local modification that works for you then this definitely seems like a reasonable way to go about it, but for an upstream PR minimising the number of files changed and avoiding a larger refactor in favour of the smallest working change is not a priority.

I would rather make sure that it is fixed robustly and in the way that I consider most appropriate, which is to handle the ordering during replay rather than capture. This certainly is up to subjectivity, as it sounds like you want something different from what I think is best - I would definitely not want to have several independent event browser tabs for different queue submissions. I do agree that GPU order is useful but I don't agree that CPU submission order is irrelevant.

The bottom line is that for me I don't agree that a solution which I would bluntly (but to be clear - not entirely critically) term a hack is strictly better than no solution at all. I don't think this is a case of letting perfect be the enemy of good, in my view this is about not taking on further technical debt by doing a partial fix.

As a note just to clarify: this code to access the fences can easily break _today_ without any other code changes. That's why I'm saying that poking into serialisation internals is not going to be OK for a robust solution.

**Comment by GertyP** [comment]
> I would definitely not want to have several independent event browser tabs for different queue submissions.

I'd like to better understand your thinking on this.  Viewing a linear sequence of events _by queue_ (as opposed to a mix of stuff happening across multiple queues, interleaved into a single list of events) seems like an obvious improvement.  I am repeating myself but because it focuses on sequence of events for each queue it sidesteps the arguably misleading concern of how to interleave multiple concurrent sequences into a single stream.  A per-queue-centric view of events is a simple and logical way to reason about and understand things, while having the additional benefit of giving users a 'filter by queue' view of events to really clean things up.

Even if we came up with a way to interleave and order concurrent events into a single sequence, doing so would present to the user  quite an explicit sequence of events that doesn't necessarily reflect the ordering that truly determines the results; the GPU execution order.  E.g. -
```
Compute Q0:  Wait fence A, Dispatch U, Dispatch V, Dispatch W
Compute Q1:  Wait fence A, Dispatch X, Dispatch Y, Dispatch Z
Gfx Q: Signal A
```
Choosing to focus on CPU submission order, we might linearise events to something like -
```
Wait A, U, V, W, Wait A, X, Y, Z, Signal A
```
Which is obviously pretty unhelpful/misleading whatever way you look at it.
Even if we wanted a more sensible execution order _but still in a single linear sequence_, we might choose to show it as -
```
Signal A, Wait A, U, V, W, Wait A, X, Y, Z
```
which is a bit better but still strongly suggests that U, V, and W all are done before X, Y, and Z, which is dangerous and quite likely misleading.  Better still to show the 3 separate streams -
```
Signal A
Wait A, U, V, W
Wait A, X, Y, Z
```
and leave the relative ordering between U, V, W, and X, Y, Z as rightfully ambiguous.

> I do agree that GPU order is useful but I don't agree that CPU submission order is irrelevant.

I don't think I said CPU submission order is irrelevant... but it really is a very distant second to GPU/execution ordering.

I won't press the point any more but I'm coming from the point of view of this approach being tried and tested (and intuitive) in Pix and other (console) GPU analysis tools do this and they work well and display things simply and cleanly.  I'm not claiming it's the only way to go but just because I'm struggling to understand a couple of things you seem quite set on (focusing on CPU over GPU execution ordering and interleaving multiple concurrent streams into a single event list) I do wonder if I've misunderstood something or can get you to provide a bit more detail whenever you can find the time.

**Comment by baldurk** [comment]
To clarify - I'm certainly not focussing on CPU over GPU ordering. I agreed with you initially that we'd want to prioritise the GPU execution order for display, and I do think it is the most useful ordering. You're right that you didn't say the CPU order is irrelevant but the issue is that the current changes effectively treat it as such by losing that information. I agree the CPU order has limited value but it does have _some_ value. If we do the re-ordering into GPU execution order at capture time as if the application had not used wait-before-signal at all, then we have lost the information about what order things were really submitted in. If we do the re-ordering at replay time then we have the original CPU ordering _and_ we can still reorder things into GPU execution order.

Regarding the UI I get that you don't feel the same way - that's fair because it's a subjective determination - but I do think that having a single linearised view of events in the event browser is the best path. In your example, I would want to see the `Signal A, Wait A, U, V, W, Wait A, X, Y, Z` as the displayed order. I don't agree that it necessarily causes confusion as even at the command buffer level in modern APIs the linear list of events does not necessarily imply a strict order either. Similarly the act of stepping through a frame in a selected set of events implicitly does produce an ordering and I think it's best to have that consistently reflected.

It would certainly be useful to be able to visualise the parallelism between queues and the different orderings, but I think the event browser is not the place to do that and having a single list of events for browsing the frame is better than having multiple lists. I'm not familiar with console tools and I've only used PIX briefly, but I don't necessarily agree with every design decision that other tools have made and I do my best to make what I believe are the best decisions for RenderDoc. FWIW I checked how nsight treats this, and it seems to list a linear set of events in GPU execution order as I described above.

(Side note, this is getting a little off track - I do appreciate that the UI display of events is a related issue but it's not completely the same thing as how we handle wait-before-signal in general. I don't mind discussing it a bit here as a related issue but I don't want to move the goalposts too far and it doesn't change my view of the original changes in this PR.)

From a practical perspective, as I say the UI approach is _mostly_ orthogonal to how wait-before-signal is properly replayed. The exceptions being that unless we preserve the CPU submission order we can't ever display that information because it's lost, and that if we don't change the UI at all we need to decide somehow how to produce a linear sequence of events for the event browser.

For that reason I would still want to handle this re-ordering implementation on the replay side, and I would want to handle it robustly by tracking all fence values strictly in order to do so, without fallbacks for forward-progress unblocking. In terms of the UI this would then effectively produce the same result as your current changes - the event browser would still display a single linear list of events but now in GPU execution order. After that it would be a further conversation or set of changes to expand or change how the UI displays parallel queue work.

**Comment by GertyP** [comment]
> then we have lost the information about what order things were really submitted in

This got me thinking:  I genuinely can't remember ever wanting or needing to consider or even care about CPU submission order of GPU work when using any graphics analysis tools; only ever "what work is the GPU sucking up and what is the stream of work executed on each of the concurrent units/queues that produce the results we're actually seeing as output?".  Again, I wouldn't dare to say the CPU sub. order is of no value ... only I struggle to think of when I'd need to see anything other than GPU execution sequences.

> It would certainly be useful to be able to visualise the parallelism between queues and the different orderings, but I think the event browser is not the place to do that

Agreed.  Pix does this well with a separate timeline view that shows dispatches/draws/copy/flushing/etc work along each line, for each gpu queue and, most helpfully, with a nice little 'signal-to-wait' arrow that points from the signalling queue timeline point to the waiting queue's timeline point.

> and having a single list of events for browsing the frame is better than having multiple lists

I understand but, personally, I disagree.  Having used Pix's (and other console analysis tools) with per-queue view of events for many years, to see a clean view of what each queue is fed, I find it hard to switch back to a more messy, interleaved, single list of work from many different gpu queues.  If I know I've got some copying/downsampling/shadowmap dispatches being done on a particular async compute, it's nice to simply view that queue's events without all the mess from everything else going through all the other queues.  It's also clearer to see push/pop-style marker events that bracket blocks of work sent to a particular queue _only_ for that queue's event list, whereas a single merged/interleaved sequence can easily become a noisy, overlapping mess of multiple begin/end marker events.

> and I would want to handle it robustly by tracking all fence values strictly in order to do so

Yes, that would be ideal;  a 'd3d12_fence_wrap.cpp/h' to record and pass through creation and CPU-side `ID3D12Fence::Signal` events... Still, it's feels unfortunate we currently have a 'just plain wrong' multi-queue execution replay when there's a more useful, more respecting (but perhaps not perfect, but no worse), low surface-area execution replay to hand.

... Anyway, I just wanted to say that I very much appreciate the time you've taken and the way you've gone about chatting through your thoughts on this. 👍 

I can see this is the end of the line for this draft PR 😞 .  Does it make any difference whether you close it or I do?
